DW-158: Order of execution in Function Overloading (#1649)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/dataweave-functions.adoc

"(123) 456-7890"
----

=== Overloading Functions

> DataWeave allows you to create multiple functions with the same name but different parameters. This is useful for defining different behaviors based on the arguments of a function call.

Reviewer: ```suggestion
DataWeave enables you to create multiple functions with the same name but different parameters. This feature is useful for defining different behaviors based on the arguments of a function call.
```

> The difference in parameters could be based on the number of parameters or the types of those parameters. To understand more about defining type constraints on function parameters you can check the xref:dataweave-type-system.adoc[type system] docs.

Reviewer: ```suggestion
Parameters in overloaded functions differ in number or type. To understand more about defining type constraints on function parameters, see xref:dataweave-type-system.adoc[Type System].
```

Author: Maybe "Parameters in overloaded functions have to differ in number or type." just to be a little bit more clear.??

Reviewer: good point, done

> DataWeave uses the first function it finds that accepts the arguments of the function call, in the order that they are declared in the script.

Reviewer: ```suggestion
DataWeave uses the first function it finds that accepts the arguments of the function call, based on the order in which the functions are declared in the script.
```

.Example: Function that Uses Function Overloading
[source,dataweave,linenums]
----
%dw 2.0
output application/json

fun toUpper(a: String) = upper(a)
fun toUpper(a: Any) = null
fun toUpper(a: String, b: Number) = upper(a) ++ b as String
---
toUpper("hi!")
----

> The argument of the function call `toUpper("hi!")` matches with the types String and Any (so it could be called with the first two definitions) but DataWeave will execute the function `fun toUpper(a: String) = upper(a)` because is defined before the one that uses the type Any.

Reviewer: ```suggestion
The argument of the function call `toUpper("hi!")` matches the types String and Any, so it is possible to call the function with the first two definitions. However, DataWeave executes the function `fun toUpper(a: String) = upper(a)` because that function is defined before the one that uses the type Any.
```
